File: c/tcp_server_sum/tcp_server_sum.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#include "constants.h"

#define BUFFER_SIZE 300
#define SERVER_PORT "9999"

int main(void) {
    struct sockaddr_storage client_addr;
    int err, new_sock, sock;
    unsigned int addr_len;
    unsigned long i, f, n, num, sum;
    unsigned char bt;
    char cli_ip_text[BUFFER_SIZE], cli_port_text[BUFFER_SIZE];
    struct addrinfo req, *list;

    bzero((char *)&req, sizeof(req));
    req.ai_family = AF_INET6; // allow both IPv4 and IPv6 clients to use this server
    req.ai_socktype = SOCK_STREAM;
    req.ai_flags = AI_PASSIVE; // local address
    err = getaddrinfo(NULL, SERVER_PORT, &req, &list);
    if(err) {
        printf("Failed to get local address, error: %s\n",gai_strerror(err));
        exit(INVALID_LOCAL_ADDRESS);
    }

    sock = socket(list->ai_family, list->ai_socktype, list->ai_protocol);
    if(sock < 0) {
        perror("Failed to open socket");
        freeaddrinfo(list);
        exit(OPEN_SOCKET_FAILED);
    }

    if(bind(sock, (struct sockaddr *)list->ai_addr, list->ai_addrlen) < 0) {
        perror("Bind failed");
        close(sock);
        freeaddrinfo(list);
        exit(BIND_SOCKET_FAILED);
    }
    freeaddrinfo(list);

    listen(sock, SOMAXCONN);
    puts("Accepting TCP connections (both over IPv6 or IPv4). Use CTRL+C to terminate the server");
    addr_len = sizeof(client_addr);

    while(1) {
        new_sock = accept(sock, (struct sockaddr *)&client_addr, &addr_len);
        if(fork() == 0)
        {
            close(sock);
            getnameinfo((struct sockaddr *)&client_addr, addr_len, cli_ip_text, BUFFER_SIZE,
                        cli_port_text, BUFFER_SIZE, NI_NUMERICHOST|NI_NUMERICSERV);
            printf("New connection from %s, port number %s\n", cli_ip_text, cli_port_text);
            do {
                sum = 0;
                do {
                    num = 0;
                    f = 1;
                    for(i = 0; i < 4; i++) {
                        read(new_sock, &bt, 1);
                        num = num + bt * f;
                        f = 256 * f;
                    }
                    sum = sum + num;
                }
                while(num > 0);

                n = sum;
                for(i = 0; i < 4; i++) {
                    bt = n % 256;
                    write(new_sock, &bt, 1);
                    n = n / 256;
                }
            }
            while(sum > 0);
            close(new_sock);

            printf("Connection from %s, port number %s closed\n", cli_ip_text, cli_port_text);
            exit(0);
        }
        close(new_sock);

    }
    close(sock);

    return 0;
}

File: java/tcp_server_sum/TcpServerSum.java
import java.io.*;
import java.net.*;

/**
 * Represents a TCP server that handles sums.
 */
class TcpServerSum {

    public static final int SERVER_PORT = 9999;

    public static final int OPEN_SOCKET_FAILED = -1;

    /**
     * Starts the TCP server.
     *
     * @param args arguments from command line
     * @throws IOException input/output exception
     */
    public static void main(String args[]) throws IOException {
        ServerSocket sock = null;
        Socket cliSock;

        try {
            sock = new ServerSocket(SERVER_PORT);
        } catch (IOException ex) {
            System.out.println("Failed to open server socket");
            System.exit(OPEN_SOCKET_FAILED);
        }

        System.out.println("Server is running, waiting for connections (press \"CTRL + c\" to quit).");
        while (true) {
            cliSock = sock.accept();
            Thread clientThread = new Thread(new TcpServerSumThread(cliSock));
            clientThread.start();
        }
    }
}

/**
 * Thread to manage a TCP connection.
 */
class TcpServerSumThread implements Runnable {
    private Socket s;
    private DataOutputStream sOut;
    private DataInputStream sIn;

    public TcpServerSumThread(Socket clientSocket) {
        s = clientSocket;
    }

    @Override
    public void run() {
        long f, i, num, sum;
        InetAddress clientIP;
        clientIP = s.getInetAddress();
        System.out.println("New client connection from " + clientIP.getHostAddress()
                + ", port number " + s.getPort());
        try {
            sOut = new DataOutputStream(s.getOutputStream());
            sIn = new DataInputStream(s.getInputStream());
            do {
                sum = 0;
                do {
                    num = 0;
                    f = 1;
                    for (i = 0; i < 4; i++) {
                        num = num + f * sIn.read();
                        f = f * 256;
                    }
                    sum = sum + num;
                }
                while (num > 0);
                num = sum;
                for (i = 0; i < 4; i++) {
                    sOut.write((byte) (num % 256));
                    num = num / 256;
                }
            }
            while (sum > 0);
            System.out.println("Client " + clientIP.getHostAddress()
                    + ", port number: " + s.getPort() + " disconnected");
            s.close();
        } catch (IOException ex) {
            System.out.println("IOException");
        }
    }
}
